### PR TITLE
BENCH-264: Tags

### PR DESCRIPTION
**Pull request by ORybak5**
Tags introduction to our API ecosystem.

### CODE REVIEW

File: AnswerKingAPI.yml [hunk 1]
```diff
@@ -2,15 +2,230 @@ openapi: 3.0.1
...
+    description: Create and manage customer orders.
+  - name: Tags
+    description: Create and manage tags.
 paths:
+  /api/tags:
+    post:
+      tags:
+        - Tags
+      summary: Create Tag.
+      operationId: craeteTag
```
Comment (bethcryer) [inline AnswerKingAPI.yml]: typo :P

File: AnswerKingAPI.yml [hunk 2]
```diff
@@ -2,15 +2,230 @@ openapi: 3.0.1
...
+        '400':
+          description: Invalid parameters are provided.
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/ProblemDetails'
+    get:
+      tags:
+        - Tags
+      summary:  List All tags.
```
Comment (bethcryer) [inline AnswerKingAPI.yml]: is there a reason for the double-spaces here? they're showing up in a few places

File: AnswerKingAPI.yml [hunk 3]
```diff
@@ -2,15 +2,230 @@ openapi: 3.0.1
...
+      operationId: listAllTags
+      responses:
+        '200':
+          description:  OK
```
Comment (bethcryer) [inline AnswerKingAPI.yml]: another double-space

File: AnswerKingAPI.yml [hunk 4]
```diff
@@ -2,15 +2,230 @@ openapi: 3.0.1
...
+            application/json:
+              schema:
+                $ref: '#/components/schemas/ProblemDetails'
+    put:
+      tags:
+        - Tags
+      summary:  Assign tag to products.
+      operationId:  assignTagToProducts
+      requestBody:
+        description: ''
```
Comment (bethcryer) [inline AnswerKingAPI.yml]: Does this one need an empty description? could probably remove if there's no description for it

**Review by bethcryer — COMMENTED** [review]
Structure of the contract itself looks fine, just a few nitpicks I noticed

**Review by milkshakeuk — APPROVED** [review]

File: AnswerKingAPI.yml [hunk 5]
```diff
@@ -2,15 +2,230 @@ openapi: 3.0.1
...
+              schema:
+                $ref: '#/components/schemas/Tag.Response'
+        '400':
+          description: Invalid parameters are provided.
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/ProblemDetails'
+        '404':
+          description: Tag with the given id does not exist
```
Comment (AndyFlintAnswerDigital) [inline AnswerKingAPI.yml:80]: Description is missing a full stop.

File: AnswerKingAPI.yml [hunk 6]
```diff
@@ -2,15 +2,230 @@ openapi: 3.0.1
...
+    parameters:
+      - schema:
+          type: string
+        name: id
+        in: path
+        required: true
+    get:
+      tags:
+        - Tags
+      summary:  List  Products For Tag.
```
Comment (AndyFlintAnswerDigital) [inline AnswerKingAPI.yml]: Double space in the middle of the summary.

File: AnswerKingAPI.yml [hunk 7]
```diff
@@ -2,15 +2,230 @@ openapi: 3.0.1
...
+      requestBody:
+        description: ''
+        content:
+          application/json:
+            schema:
+              type: array
+              items:
+                type: integer
+              minItems: 1
+      responses:
```
Comment (AndyFlintAnswerDigital) [inline AnswerKingAPI.yml:190]: Could we also introduce a 410 response if we try to apply a tag to a retired product?

File: AnswerKingAPI.yml [hunk 8]
```diff
@@ -2,15 +2,230 @@ openapi: 3.0.1
...
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/ProblemDetails'
+    delete:
+      tags:
+        - Tags
+      summary: Remove Tag From Products.
+      operationId:  removeTagFromProducts
+      responses:
```
Comment (AndyFlintAnswerDigital) [inline AnswerKingAPI.yml:210]: I feel like this should also have a 404 response for if we try to remove a tag which isn't present on a product, though I appreciate that might be a case for a 400? Happy to discuss.

File: AnswerKingAPI.yml [hunk 9]
```diff
@@ -2,15 +2,230 @@ openapi: 3.0.1
...
+            application/json:
+              schema:
+                $ref: '#/components/schemas/ProblemDetails'
+        '404':
+          description: Tag with the given id does not exist.
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/ProblemDetails'
+        '410':
```
Comment (AndyFlintAnswerDigital) [inline AnswerKingAPI.yml:134]: There seems to be an inconsistency here with the deletion of tags. Adding the possibility of a 410 response implies that we can retire tags, however, I'm not seeing 410 responses added to our other endpoints (notably modifyTag, viewProductsForTag and assignTagToProducts, as well as possibly removeTagFromProducts if the product or tag is retired)

I've started a discussion on the bench-king-senior-dev-board Slack channelabout this.

**Review by AndyFlintAnswerDigital — CHANGES_REQUESTED** [review]

File: AnswerKingAPI.yml [hunk 10]
```diff
@@ -373,11 +600,17 @@ paths:
...
               schema:
                 $ref: '#/components/schemas/Inventory.Product.Response'
+        '400':
+          description: Invalid parameter is provided.
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/ProblemDetails'
         '404':
           description: When the product with the given id does not exist
```
Comment (MichaelOSullivanAnswer) [inline AnswerKingAPI.yml:615]: Missed one of the 'When' removals

File: AnswerKingAPI.yml (hunk 1 above)
Comment (ORybak5) [reply AnswerKingAPI.yml]: will fix it

File: AnswerKingAPI.yml (hunk 2 above)
Comment (ORybak5) [reply AnswerKingAPI.yml]: no reason, will be fixed.

File: AnswerKingAPI.yml (hunk 6 above)
Comment (ORybak5) [reply AnswerKingAPI.yml]: will be fixed

File: AnswerKingAPI.yml (hunk 9 above)
Comment (ORybak5) [reply AnswerKingAPI.yml:134]: as responded in my comment in #bench-king-senior-dev-board group, I'd delete it for now until we discuss the whole "Retire" concept for all the endpoints.

File: AnswerKingAPI.yml (hunk 4 above)
Comment (ORybak5) [reply AnswerKingAPI.yml]: I deleted it

**Review by MichaelOSullivanAnswer — APPROVED** [review]

**Review by bethcryer — APPROVED** [review]

**Pull request merged**